File: mfq/server/api/routes/sessions.py
"""Session, response, and generation-preset routes."""

from __future__ import annotations

from typing import Annotated, Any
from uuid import UUID

from fastapi import APIRouter, Query, Response
from fastapi.responses import StreamingResponse

from mfq.server.api.dependencies import ServiceDependency
from mfq.server.api.routes import ERROR_RESPONSES
from mfq.server.protocol.models import (
    AppendMessageRequest,
    AppendMessageResult,
    CreateGenerationPresetRequest,
    CreateResponseRequest,
    CreateSessionRequest,
    ForkSessionRequest,
    GenerationPresetList,
    GenerationPresetResource,
    MessageList,
    ResponseList,
    ResponseResource,
    RewindSessionRequest,
    SessionArchive,
    SessionImportResult,
    SessionList,
    SessionResource,
    UpdateGenerationPresetRequest,
    UpdateSessionRequest,
)

initial_router = APIRouter()
router = APIRouter()


@initial_router.post(
    "/api/v1/sessions",
    response_model=SessionResource,
    status_code=201,
    responses=ERROR_RESPONSES,
    tags=["sessions"],
)
async def create_session(service: ServiceDependency, body: CreateSessionRequest) -> SessionResource:
    return await service.create_session(body)


@initial_router.post(
    "/api/v1/sessions/import",
    response_model=SessionImportResult,
    status_code=201,
    responses=ERROR_RESPONSES,
    tags=["sessions"],
)
async def import_session(service: ServiceDependency, body: SessionArchive) -> SessionImportResult:
    return await service.import_session(body)


@initial_router.post(
    "/api/v1/presets",
    response_model=GenerationPresetResource,
    status_code=201,
    responses=ERROR_RESPONSES,
    tags=["presets"],
)
async def create_generation_preset(
    service: ServiceDependency,
    body: CreateGenerationPresetRequest,
) -> GenerationPresetResource:
    return await service.create_generation_preset(body)


@initial_router.get(
    "/api/v1/presets",
    response_model=GenerationPresetList,
    responses=ERROR_RESPONSES,
    tags=["presets"],
)
async def list_generation_presets(
    service: ServiceDependency,
) -> GenerationPresetList:
    return await service.list_generation_presets()


@initial_router.put(
    "/api/v1/presets/{preset_id}",
    response_model=GenerationPresetResource,
    responses=ERROR_RESPONSES,
    tags=["presets"],
)
async def update_generation_preset(
    service: ServiceDependency, preset_id: UUID, body: UpdateGenerationPresetRequest
) -> GenerationPresetResource:
    return await service.update_generation_preset(preset_id, body)


@initial_router.delete(
    "/api/v1/presets/{preset_id}",
    status_code=204,
    responses=ERROR_RESPONSES,
    tags=["presets"],
)
async def delete_generation_preset(service: ServiceDependency, preset_id: UUID) -> Response:
    await service.delete_generation_preset(preset_id)
    return Response(status_code=204)


@router.get(
    "/api/v1/sessions",
    response_model=SessionList,
    responses=ERROR_RESPONSES,
    tags=["sessions"],
)
async def list_sessions(
    service: ServiceDependency,
    limit: Annotated[int, Query(ge=1, le=200)] = 50,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> SessionList:
    return await service.list_sessions(limit=limit, offset=offset)


@router.get(
    "/api/v1/sessions/{session_id}",
    response_model=SessionResource,
    responses=ERROR_RESPONSES,
    tags=["sessions"],
)
async def get_session(service: ServiceDependency, session_id: UUID) -> SessionResource:
    return await service.get_session(session_id)


@router.patch(
    "/api/v1/sessions/{session_id}",
    response_model=SessionResource,
    responses=ERROR_RESPONSES,
    tags=["sessions"],
)
async def update_session(
    service: ServiceDependency,
    session_id: UUID,
    body: UpdateSessionRequest,
) -> SessionResource:
    return await service.update_session(session_id, body)


@router.get(
    "/api/v1/sessions/{session_id}/messages",
    response_model=MessageList,
    responses=ERROR_RESPONSES,
    tags=["sessions"],
)
async def list_messages(service: ServiceDependency, session_id: UUID) -> MessageList:
    return await service.list_messages(session_id)


@router.get(
    "/api/v1/sessions/{session_id}/export",
    response_model=SessionArchive,
    responses=ERROR_RESPONSES,
    tags=["sessions"],
)
async def export_session(service: ServiceDependency, session_id: UUID) -> SessionArchive:
    return await service.export_session(session_id)


@router.post(
    "/api/v1/sessions/{session_id}/messages",
    response_model=AppendMessageResult,
    status_code=201,
    responses=ERROR_RESPONSES,
    tags=["sessions"],
)
async def append_message(
    service: ServiceDependency,
    session_id: UUID,
    body: AppendMessageRequest,
) -> AppendMessageResult:
    return await service.append_message(session_id, body)


@router.get(
    "/api/v1/sessions/{session_id}/responses",
    response_model=ResponseList,
    responses=ERROR_RESPONSES,
    tags=["sessions"],
)
async def list_responses(
    service: ServiceDependency,
    session_id: UUID,
    limit: Annotated[int, Query(ge=1, le=1000)] = 200,
) -> ResponseList:
    return await service.list_responses(session_id, limit=limit)


@router.post(
    "/api/v1/sessions/{session_id}/responses",
    response_model=ResponseResource,
    responses={
        **ERROR_RESPONSES,
        200: {
            "description": "Completed response or an SSE event stream",
            "content": {
                "application/json": {"schema": {"$ref": "#/components/schemas/ResponseResource"}},
                "text/event-stream": {"schema": {"type": "string"}},
            },
        },
    },
    tags=["sessions"],
)
async def create_response(
    service: ServiceDependency, session_id: UUID, body: CreateResponseRequest
) -> Any:
    daemon = service
    prepared = await daemon.prepare_response(session_id, body)
    if body.stream:
        return StreamingResponse(
            daemon.stream_response(prepared),
            media_type="text/event-stream",
            headers={
                "Cache-Control": "no-cache",
                "X-Accel-Buffering": "no",
            },
        )
    return await daemon.collect_response(prepared)


@router.post(
    "/api/v1/sessions/{session_id}/responses/cancel",
    response_model=ResponseResource,
    responses=ERROR_RESPONSES,
    tags=["sessions"],
)
async def cancel_response(service: ServiceDependency, session_id: UUID) -> ResponseResource:
    return await service.cancel_response(session_id)


@router.post(
    "/api/v1/sessions/{session_id}/fork",
    response_model=SessionResource,
    status_code=201,
    responses=ERROR_RESPONSES,
    tags=["sessions"],
)
async def fork_session(
    service: ServiceDependency, session_id: UUID, body: ForkSessionRequest
) -> SessionResource:
    return await service.fork_session(session_id, body)


@router.post(
    "/api/v1/sessions/{session_id}/rewind",
    response_model=SessionResource,
    responses=ERROR_RESPONSES,
    tags=["sessions"],
)
async def rewind_session(
    service: ServiceDependency,
    session_id: UUID,
    body: RewindSessionRequest,
) -> SessionResource:
    return await service.rewind_session(session_id, body)


@router.delete(
    "/api/v1/sessions/{session_id}",
    status_code=204,
    responses=ERROR_RESPONSES,
    tags=["sessions"],
)
async def delete_session(service: ServiceDependency, session_id: UUID) -> Response:
    await service.delete_session(session_id)
    return Response(status_code=204)

File: MFQStudio/src/shared/api/resources/sessions.ts
/** 封装 sessions 领域资源请求，不保存组件状态。 */
import type {
  SessionMode,
  MessageRole,
  ContentPart,
  Session,
  SessionArchive,
  Message,
  ResponseResource,
} from '../types';
import { request, apiUrl, errorFromResponse, authorizedHeaders } from '../client';

export const sessionsApi = {
  /** 获取最近的会话列表，供侧栏与会话选择使用。 */
  async listSessions(): Promise<Session[]> {
    return (await request<{ data: Session[] }>('/api/v1/sessions?limit=200')).data;
  },

  /** 按模型和交互模式创建会话，返回服务端分配的标识与版本。 */
  createSession(
    model: string,
    mode: SessionMode,
    title?: string,
    metadata?: Record<string, unknown>,
  ): Promise<Session> {
    return request('/api/v1/sessions', {
      method: 'POST',
      body: JSON.stringify({ model, mode, title, metadata: metadata ?? {} }),
    });
  },

  /** 读取会话的权威状态和版本号，可随页面或请求取消。 */
  getSession(id: string, signal?: AbortSignal): Promise<Session> {
    return request(`/api/v1/sessions/${id}`, { signal });
  },

  /** 更新会话标题、模式或元数据，不修改历史消息。 */
  updateSession(
    id: string,
    update: { title?: string | null; mode?: SessionMode; metadata?: Record<string, unknown> },
  ): Promise<Session> {
    return request(`/api/v1/sessions/${id}`, {
      method: 'PATCH',
      body: JSON.stringify(update),
    });
  },

  /** 获取持久化消息，供首次加载及生成完成后的最终同步使用。 */
  async listMessages(id: string, signal?: AbortSignal): Promise<Message[]> {
    return (await request<{ data: Message[] }>(`/api/v1/sessions/${id}/messages`, { signal })).data;
  },

  /** 获取响应状态与性能记录，用于核对请求标识和最终消息。 */
  async listResponses(id: string, signal?: AbortSignal): Promise<ResponseResource[]> {
    return (
      await request<{ data: ResponseResource[] }>(`/api/v1/sessions/${id}/responses?limit=1000`, {
        signal,
      })
    ).data;
  },

  /** 请求服务端取消该会话正在运行的生成，调用方负责确认最终状态。 */
  cancelResponse(id: string, signal?: AbortSignal): Promise<ResponseResource> {
    return request(`/api/v1/sessions/${id}/responses/cancel`, { method: 'POST', signal });
  },

  /** 使用预期版本追加消息，版本冲突由服务端返回错误。 */
  appendMessage(
    id: string,
    expectedRevision: number,
    role: MessageRole,
    parts: ContentPart[],
  ): Promise<{ session: Session; message: Message }> {
    return request(`/api/v1/sessions/${id}/messages`, {
      method: 'POST',
      body: JSON.stringify({ expected_revision: expectedRevision, role, parts }),
    });
  },

  /** 从指定历史位置派生会话，可同时指定新的模型。 */
  forkSession(
    id: string,
    atMessageId: string | null,
    includeMessage = true,
    title?: string | null,
    model?: string,
  ): Promise<Session> {
    return request(`/api/v1/sessions/${id}/fork`, {
      method: 'POST',
      body: JSON.stringify({
        at_message_id: atMessageId,
        include_message: includeMessage,
        title,
        model,
      }),
    });
  },

  /** 按版本回退会话历史，供编辑和重新生成操作使用。 */
  rewindSession(
    id: string,
    expectedRevision: number,
    atMessageId: string,
    includeMessage = true,
  ): Promise<Session> {
    return request(`/api/v1/sessions/${id}/rewind`, {
      method: 'POST',
      body: JSON.stringify({
        expected_revision: expectedRevision,
        at_message_id: atMessageId,
        include_message: includeMessage,
      }),
    });
  },

  /** 删除指定服务端会话，成功后由调用方清理界面状态。 */
  async deleteSession(id: string): Promise<void> {
    const response = await fetch(apiUrl(`/api/v1/sessions/${id}`), {
      method: 'DELETE',
      headers: authorizedHeaders(),
    });
    if (!response.ok) throw await errorFromResponse(response);
  },

  /** 导出会话及媒体归档，供用户下载或备份。 */
  exportSession(id: string): Promise<SessionArchive> {
    return request(`/api/v1/sessions/${id}/export`);
  },

  /** 导入会话归档并返回新会话及媒体导入数量。 */
  importSession(archive: SessionArchive): Promise<{
    session: Session;
    messages_imported: number;
    media_imported: number;
  }> {
    return request('/api/v1/sessions/import', {
      method: 'POST',
      body: JSON.stringify(archive),
    });
  },
};
